designs/NuGetExtensibility/INuGetProjectServices.md

```diff
@@ -0,0 +1,122 @@
+
+See [NuGet extensibility services](NuGetExtensibilityServices.md) for more details.
+
+## Who are the customers
+
+Visual Studio component and extension developers.
+
+## Goals
+
+* Can be called on any thread (ideally background thread, but work if called on UI thread).
+* Quickly return incomplete task to unblock thread.
+* Works with Visual Studio Codespaces, and hopefully Live Share.
+
+## Non-Goals
+
+* This will not be a 1:1 replacement for any of NuGet's existing `IVs` interfaces.
+
+## Solution
+
+```cs
+/// <summary>Service to interact with projects in a solution</summary>
+public interface INuGetProjectServices
+{
+    /// <summary>Gets the list of packages installed in a project.</summmry>
+    /// <param name="project">Path and filename to project file.</param>
+    /// <param name="options">Options to use. If null, uses GetInstalledPakcagesOptions.Default.</param>
+    /// <param name="cancellationToken">Cancellation token.</param>
+    /// <returns>
+    /// The list of packages in the project.
+    /// If the project does not exist in the solution, or the project is unloaded, throws ArgumentException.
+    /// </return>
+    Task<GetInstalledPackagesResult> GetInstalledPackagesAsync(string project, CancellationToken cancellationToken);
+}
+
+/// <summary>Result of a call to INuGetProjectServices.GetInstalledPackagesAsync</summary>
+public sealed class GetInstalledPackagesResult
+{
+    /// <summary>The status of the result</summary>
+    public GetInstalledPackageResultStatus Status { get; }
+
+    /// <summary>List of packages in the project</summary>
+    /// <remarks></remarks>
+    public IReadOnlyCollection<InstalledPackage> Packages { get; }
+}
+
+/// <summary>The status of the result</summary>
+public enum  GetInstalledPackageResultStatus
+{
+    /// <summary>Successful</summary>
+    Successful,
+
+    /// <summary>The project is not yet ready</summary>
+    /// <remarks>This typically happens shortly after the project is loaded, but the project system has not yet informed NuGet about package references</remarks>
+    ProjectNotReady,
+
+    /// <summary>Package information could not be retrieved because the project is in an invalid state</summary>
+    /// <remarks>If a project has an invalid target framework value, or a package reference has a version value, NuGet may be unable to generate basic project information, such as requested packages.</remarks>
+    ProjectInvalid
+}
+
+/// <summary>Basic information about a package</summary>
+public sealed class InstalledPackage
+{
+    /// <summary>The package id.</summary>
+    public string Id { get; }
+
+    /// <summary>The package version</summary>
+    /// <remarks>
+    /// If the project uses packages.config, this will be the installed package version.
+    /// If the project uses PackageReference, this would ideally be the resolved package version, but may be the requested package version.
+    /// </remarks>
+    public string Version { get; }
+
```

Comment:
Should we expose a data model similar of what represents an installed package in project.assets.json?

Reply:
I don't believe so. The assets file is NuGet's "internal" representation of whatever we need to know (well, the .NET SDK knows about the format too). But callers of this extensibility API shouldn't need to know. Plus it would mean that each time we change the assets file we'd need to create new APIs here, and maintaining backwards compatibility with the older APIs becomes significant technical debt.